src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -0,0 +1,113 @@
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.regexp;
+
+import org.joni.Regex;
+import org.truffleruby.Layouts;
+import org.truffleruby.builtins.CoreClass;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.core.regexp.RegexpNodes.ToSNode;
+import org.truffleruby.core.rope.CodeRange;
+import org.truffleruby.core.rope.Rope;
+import org.truffleruby.core.string.StringNodes.StringAppendPrimitiveNode;
+import org.truffleruby.core.string.StringOperations;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+
+@CoreClass("Truffle::RegexpOperations")
+public class TruffleRegexpNodes {
+
+    @CoreMethod(names = "union", onSingleton = true, required = 2, rest = true)
+    public static abstract class RegexpUnionNode extends CoreMethodArrayArgumentsNode {
+
+        @Child StringAppendPrimitiveNode appendNode = StringAppendPrimitiveNode.create();
+        @Child ToSNode toSNode = ToSNode.create();
+
+        @Specialization(guards = "argsMatch(cachedArgs, args)")
+        public DynamicObject executeFastUnion(VirtualFrame frame, DynamicObject str, DynamicObject sep, Object[] args,
+                @Cached(value = "args", dimensions = 1) Object[] cachedArgs,
+                @Cached("buildUnion(frame, str, sep, args)") DynamicObject union) {
+            return union;
+        }
+
+        @Specialization
+        public DynamicObject executeSlowUnion(VirtualFrame frame, DynamicObject str, DynamicObject sep, Object[] args) {
+            return buildUnion(frame, str, sep, args);
+        }
+
+        public DynamicObject buildUnion(VirtualFrame frame, DynamicObject str, DynamicObject sep, Object[] strs) {
+
+            DynamicObject regexpString = null;
+            for (int i = 0; i < strs.length; i++) {
+                if (regexpString == null) {
+                    regexpString = str;
+                    regexpString = appendNode.executeStringAppend(str, string(frame, strs[i]));
+                } else {
+                    regexpString = appendNode.executeStringAppend(regexpString, sep);
+                    regexpString = appendNode.executeStringAppend(regexpString, string(frame, strs[i]));
+                }
+            }
+            return createRegexp(StringOperations.rope(regexpString));
+        }
+
+        public DynamicObject string(VirtualFrame frame, Object obj) {
+            if (obj instanceof Rope) {
+                Rope rope = (Rope) obj;
+                boolean isAsciiOnly = rope.getEncoding().isAsciiCompatible() && rope.getCodeRange() == CodeRange.CR_7BIT;
```

Comment:
There's no need to check the encoding here. The code range can only be `CR_7BIT` if the encoding is ASCII-compatible. There's also `Rope#isAsciiOnly`, which does the same thing.